torch/csrc/jit/codegen/onednn/graph_helper.cpp

```diff
@@ -0,0 +1,521 @@
+
+    case aten::linear: {
+      return Operator(node, opkind::MatMul)
+          .setInput(0, 1, 2)
+          .setOutput(0)
+          .setAttr("transpose_b", true);
+    }
+
+    default:
+      return makeWildcardOp(node);
+  }
+}
+
+dnnl::graph::op createLlgaOp(Node* node) {
+  return createOperator(node).llgaOp();
+}
+
+bool isSupported(Node* node) {
+  return createOperator(node).kind() != opkind::Wildcard;
+};
+
+DeviceType inferDeviceFromValue(Value* v) {
+  auto tt = v->type()->cast<TensorType>();
+  if (!tt)
+    return at::kCPU;
+  auto device = tt->device();
+  if (!device)
+    return at::kCPU;
+  return device->type();
+}
+
+DeviceType inferDevice(const std::shared_ptr<Graph>& graph) {
+  auto dt = inferDeviceFromValue(graph->inputs()[0]);
+  TORCH_CHECK(
+      std::all_of(
+          graph->inputs().begin(),
+          graph->inputs().end(),
+          [dt](Value* v) { return inferDeviceFromValue(v) == dt; }),
+      "All inputs must have the same deive type");
+  return dt;
+}
+
+dnnl::graph::engine::kind getLlgaEngineKind(DeviceType type) {
+  switch (type) {
+    case DeviceType::CPU:
+      return dnnl::graph::engine::kind::cpu;
+    default:
+      TORCH_CHECK(false, "Not support device type ", type);
+  }
+}
+
+void mayAddListConstructIntoConcatPartition(
+    Node* n,
+    OpPartitionMap& opToOwningPartition) {
+  // Since prim::ListConstruct is not visible to the LLGA,
+  // it will not be in any partition returned from partfuseritioning results.
+  // We need rewrite opToOwningPartition to make the prim::ListConstruct to be
+  // 'virtually' in the same partition with the aten::cat, so that
+  // prim::ListConstruct can be fused into the fusion group by graph fuser.
+  // We emphasize on 'virtually' because get_num_ops() for cat's partition
+  // would still return 1.
+  if (n->kind() == aten::cat && opToOwningPartition.has(n)) {
+    auto listConstrcut = n->input(0)->node();
+    auto partitionId = opToOwningPartition.get(n);
+    opToOwningPartition.add(listConstrcut, partitionId);
+  }
+}
+
+LlgaGraphHelper::LlgaGraphHelper(
+    const std::shared_ptr<Graph>& graph,
+    dnnl::graph::partition::policy policy) {
+  auto deviceType = inferDevice(graph);
+  auto engineKind = getLlgaEngineKind(deviceType);
+  dnnl::graph::graph g{engineKind};
+
+  GRAPH_DEBUG("Constructing LLGA graph");
+  // TODO: select nodes in top-level block for now
+  for (auto* node : graph->block()->nodes()) {
+    auto op = createLlgaOp(node);
+    try {
+      g.add_op(op);
+    } catch (std::exception& e) {
+      GRAPH_DEBUG(
+          "The backend failed to add node ", node->kind().toQualString());
+      g.add_op(makeWildcardOp(node).llgaOp());
+    }
+    GRAPH_DEBUG("  Added node ", node->kind().toQualString());
+
+    for (Value* input : node->inputs()) {
+      tensorIdToValue_.emplace(input->unique(), input);
+    }
+  }
+
+  GRAPH_DEBUG("Get Partitions");
+  std::vector<dnnl::graph::partition> partitions = g.get_partitions(policy);
```

Comment:
Do we need to create the partially initially before fusing ? Why not fuse and then create partitions at the end ? would this get rid of  needing to maintain the partition map ?

Reply:
Sorry, I should've added this in the comments - The oneDNN Graph library determines which ops would be in the same partition, so we need to create partitions to determine what ops ended up in the same partition. Currently, we've a 2-stage partition-building approach. 

When we first create the oneDNN Graph partitions during `CreateLlgaSubgraphs`, `GraphRewriter`'s `LlgaGraphHelper` object creates a partition map that is not maintained after the `CreateLlgaSubgraphs` pass. 
To maintain this map (and the partitions) for later use, we could probably make it a member of the graph instead, so that they could be freed after the graph would be destructed, but that'd complicate things. We welcome your suggestions  to improve this design.

Now, since we don't maintain the partition-map (or the partitions) created in the first step, we re-create them while creating fusion groups (`prim::oneDNNFusionGroup` are created with `createLlgaKernel`).